VirtoCommerce.CustomerModule.Data/Handlers/LogChangesMemberChangedEventHandler.cs

```diff
@@ -0,0 +1,120 @@
+
+namespace VirtoCommerce.CustomerModule.Data.Handlers
+{
+    /// <summary>
+    /// Represents logic for audit logging for contact changes.
+    /// </summary>
+    public class LogChangesMemberChangedEventHandler : IEventHandler<MemberChangedEvent>
+    {
+        private readonly IChangeLogService _changeLogService;
+        private readonly string[] _observedProperties;
+
+        public static string[] DefaultObservedProperties
+        {
+            get
+            {
+                var memberPropsNames = ReflectionUtility.GetPropertyNames<Contact>(
+                    x => x.Name, x => x.FirstName, x => x.LastName, x => x.MiddleName,
+                    x => x.Salutation, x => x.FullName, x => x.BirthDate);
+                return memberPropsNames.ToArray();
+            }
+        }
+
+        public LogChangesMemberChangedEventHandler(IChangeLogService changeLogService)
+            : this(changeLogService, DefaultObservedProperties)
+        {
+        }
+
+        protected LogChangesMemberChangedEventHandler(IChangeLogService changeLogService, string[] observedProperties)
+        {
+            _changeLogService = changeLogService;
+            _observedProperties = observedProperties;
+        }
+
+        public virtual Task Handle(MemberChangedEvent message)
+        {
+            //Log audit records only for contacts members
+            var operationLogs = message.ChangedEntries.Where(x => x.EntryState == EntryState.Modified && x.OldEntry is Contact)
+                                        .SelectMany(GetChangedEntryOperationLogs)
+                                        .ToArray();
+            if (operationLogs.Any())
+            {
+                _changeLogService.SaveChanges(operationLogs);
+            }
+            return Task.CompletedTask;
+        }
+
+        protected virtual IEnumerable<OperationLog> GetChangedEntryOperationLogs(GenericChangedEntry<Member> changedEntry)
+        {
+            var result = new List<string>();
+
+            var oldContact = changedEntry.OldEntry as Contact;
+            var changedContact = changedEntry.NewEntry as Contact;
+
+            var diff = Comparer.Compare(oldContact, changedContact);
+            //First, we detect changes with simple properties of the contact object.
+            var observedDifferences = diff.Join(_observedProperties, x => x.Name.ToLowerInvariant(), x => x.ToLowerInvariant(), (x, y) => x).ToArray();
+            var anonymousDiffComparer = AnonymousComparer.Create((Difference x) => string.Join(":", x.Name, x.NewValue, x.OldValue));
+            foreach (var difference in observedDifferences.Distinct(anonymousDiffComparer))
+            {
+                result.Add(string.Format(MemberResources.MemberPropertyChanged, difference.Name, difference.OldValue, difference.NewValue));
+            }
+            //Second, detect changes for contact dependencies
+            result.AddRange(GetContactDependenciesChanges(oldContact, changedContact));
```

Comment:
They are not dependencies for the contacts. Rather "dependants", but this name is also not correct, because neither of them depend on each other.